### PR TITLE
MAINT: Chain some exceptions.

### PR DESCRIPTION
**Pull request by zaak71**
This is related to issue #15986. I have modified a few files in the numpy and numpy/core folders to fit what has been requested here. This is my first PR on GitHub so if I did anything wrong please let me know. Thank you.

### CODE REVIEW

File: numpy/core/_dtype.py [hunk 1]
```diff
@@ -24,11 +24,11 @@
 def _kind_name(dtype):
     try:
         return _kind_to_stem[dtype.kind]
-    except KeyError:
+    except KeyError as e:
         raise RuntimeError(
             "internal dtype error, unknown kind {!r}"
             .format(dtype.kind)
-        )
+        ) from e
```
Comment (eric-wieser) [inline numpy/core/_dtype.py:31]: I'd updated https://github.com/numpy/numpy/issues/15986 to be a little clearer - this is a case where `from None` would be better.

File: numpy/ctypeslib.py [hunk 2]
```diff
@@ -377,10 +377,10 @@ def _ctype_from_dtype_scalar(dtype):
         dtype_native = dtype.newbyteorder('=')
         try:
             ctype = _scalar_type_map[dtype_native]
-        except KeyError:
+        except KeyError as e:
             raise NotImplementedError(
                 "Converting {!r} to a ctypes type".format(dtype)
-            )
+            ) from e
```
Comment (eric-wieser) [inline numpy/ctypeslib.py:383]: Another case where `from None` would produce a better message

**Comment by eric-wieser** [comment]
Thanks for the patch! Everything looks great with the PR, just looks like I was unclear about how to choose between the two options in the tracking issue.

File: numpy/core/records.py [hunk 3]
```diff
@@ -509,8 +509,8 @@ def __setattr__(self, attr, val):
                     return ret
         try:
             res = fielddict[attr][:2]
-        except (TypeError, KeyError):
-            raise AttributeError("record array has no attribute %s" % attr)
+        except (TypeError, KeyError) as e:
+            raise AttributeError("record array has no attribute %s" % attr) from e
```
Comment (rossbar) [inline numpy/core/records.py:513]:
```suggestion
            raise AttributeError(
                "record array has no attribute %s" % attr
            ) from e
```
Style nit: 80-char line limit

**Review by rossbar — COMMENTED** [review]
Just  a style nit. It might also be nice to convert the strings in `AttributeError`s in `numpy/core/records.py` to f-strings (this might get the one suggestion below the line limit on its own) or to use `.format`, but it's not necessary.

**Comment by zaak71** [comment]
Ok, I've implemented the feedback I've gotten. I also found some other places where I could fix the exceptions like this, should I just wait for this to be merged or should I keep working on this branch?

**Review by rossbar — APPROVED** [review]
LGTM - I'd say hold off on new modifications (i.e. adding more chained exceptions) so the review doesn't get too muddled.

**Comment by mattip** [comment]
Thanks @zaak71

**Pull request merged**